src/vmm/src/persist.rs

```diff
@@ -498,11 +541,63 @@ fn guest_memory_from_file(
 }
 
 fn guest_memory_from_uffd(
-    _mem_file_path: &Path,
-    _mem_state: &GuestMemoryState,
-    _track_dirty_pages: bool,
+    mem_uds_path: &Path,
+    mem_state: &GuestMemoryState,
+    track_dirty_pages: bool,
 ) -> std::result::Result<GuestMemoryMmap, LoadSnapshotError> {
-    unimplemented!()
+    use self::LoadSnapshotError::{
+        CreateUffdBuilder, DeserializeMemory, UdsConnection, UffdMemoryRegionsRegister, UffdSend,
+    };
+
+    let guest_memory =
+        GuestMemoryMmap::restore(None, mem_state, track_dirty_pages).map_err(DeserializeMemory)?;
+
+    let uffd = UffdBuilder::new()
+        .close_on_exec(true)
+        .non_blocking(true)
+        .create()
+        .map_err(CreateUffdBuilder)?;
+
+    let mut backend_mappings = Vec::with_capacity(guest_memory.num_regions());
+    for (mem_region, state_region) in guest_memory.iter().zip(mem_state.regions.iter()) {
+        let host_base_addr = mem_region.as_ptr();
+        let size = mem_region.size();
+
+        uffd.register(host_base_addr as _, size as _)
+            .map_err(UffdMemoryRegionsRegister)?;
+        backend_mappings.push(RegionBackendMapping {
+            base_h_va: host_base_addr as u64,
+            size,
+            offset: state_region.offset,
+        });
+    }
+
+    // This is safe to unwrap() because we control the contents of the vector
+    // (i.e RegionBackendMapping entries).
+    let backend_mappings = serde_json::to_string(&backend_mappings).unwrap();
+
+    let socket = UnixStream::connect(mem_uds_path).map_err(UdsConnection)?;
+    socket
+        .send_with_fd(
+            backend_mappings.as_bytes(),
+            // In the happy case we can close the fd since the other process has it open and is
+            // using it to serve us pages.
+            // The problem is that if other process crashes/exits, firecracker guest memory
+            // will simply revert to anon-mem behavior which would lead to silent errors and
+            // undefined behavior.
+            //
+            // If the fault serving process exits while we hold a copy of the FD as well, the
+            // uffd will still be alive but with no one to serve faults leading to guest freeze
+            // (which is an unfortunate, but explicit and defined behavior).
```

Comment:
Do we plan to address this in the future ? Is there anything we can do to handle this scenario ?

Reply:
Not really, no. What we are recommending is for the page fault handler process to make sure it kills the Firecracker process in case any error ensues while trying to handle page faults. In that respect, we are thinking of including the Firecracker pid in the UDS payload, so that the page fault handler can bring down the faulting thread with him.